server/datastore/mysql/software.go

```diff
@@ -2872,243 +2278,753 @@ last_vpp_install AS (
+				AND software_installers.platform IN (:host_compatible_platforms)
+				AND software_installers.global_or_team_id = :global_or_team_id
+			LEFT JOIN
+				vpp_apps ON software_titles.id = vpp_apps.title_id
+				AND vpp_apps.platform = :host_platform
+			LEFT JOIN
+				vpp_apps_teams ON vpp_apps.adam_id = vpp_apps_teams.adam_id
+				AND vpp_apps.platform = vpp_apps_teams.platform
+				AND vpp_apps_teams.global_or_team_id = :global_or_team_id
+			LEFT JOIN
+				host_software ON host_software.host_id = :host_id
+				AND host_software.software_id = software_titles.id
+			LEFT JOIN
+				host_software_installs ON host_software_installs.host_id = :host_id
+				AND host_software_installs.software_installer_id = software_installers.id
+				AND host_software_installs.removed = 0
+			LEFT JOIN
+				upcoming_activities AS ua_si ON ua_si.host_id = :host_id
+				AND ua_si.activity_type = 'software_install'
+			LEFT JOIN
+				software_install_upcoming_activities AS siua ON siua.upcoming_activity_id = ua_si.id
+				AND siua.software_installer_id = software_installers.id
+			LEFT JOIN
+				upcoming_activities AS ua_su ON ua_su.host_id = :host_id
+				AND ua_su.activity_type = 'software_uninstall'
+			LEFT JOIN
+				software_install_upcoming_activities AS suua ON suua.upcoming_activity_id = ua_su.id
+				AND suua.software_installer_id = software_installers.id
+			LEFT JOIN
+				host_vpp_software_installs ON host_vpp_software_installs.host_id = :host_id
+				AND host_vpp_software_installs.adam_id = vpp_apps_teams.adam_id
+				AND host_vpp_software_installs.removed = 0
+			LEFT JOIN
+				upcoming_activities AS ua_vpp ON ua_vpp.host_id = :host_id
+				AND ua_vpp.activity_type = 'vpp_app_install'
+			LEFT JOIN
+				vpp_app_upcoming_activities AS vaua ON vaua.upcoming_activity_id = ua_vpp.id
+				AND vaua.adam_id = vpp_apps_teams.adam_id
+			WHERE
+				host_software.software_id IS NULL
+				AND host_software_installs.software_installer_id IS NULL
+				AND siua.software_installer_id IS NULL
+				AND suua.software_installer_id IS NULL
+				AND host_vpp_software_installs.adam_id IS NULL
+				AND vaua.adam_id IS NULL
+				AND (software_installers.id IS NOT NULL OR vpp_apps_teams.platform = :host_platform)
+		`
+		stmtAvailable, args, err = sqlx.Named(stmtAvailable, namedArgs)
+		if err != nil {
+			return nil, nil, err
+		}
+		stmtAvailable, args, err = sqlx.In(stmtAvailable, args...)
+		if err != nil {
+			return nil, nil, ctxerr.Wrap(ctx, err, "expand IN query for list host software")
+		}
+		var availableSoftwareTitles []*hostSoftware
+		err = sqlx.SelectContext(ctx, ds.reader(ctx), &availableSoftwareTitles, stmtAvailable, args...)
+		if err != nil {
+			return nil, nil, err
+		}
+		// Clear out all the previous software titles as we are only filtering for available software
+		if opts.OnlyAvailableForInstall {
+			bySoftwareID = make(map[uint]*hostSoftware, len(availableSoftwareTitles))
+		}
```

Comment:
I don't like this approach, since we still fire off all the queries above but just discard the results. Should figure out a way to optimize here, and only fire off the available software query.